### ISSUE TITLE
the different between khmot and sort?

### ISSUE THREAD

**Issue by daiwc** [r7vme/khmot#1]

**Comment by r7vme** [comment]
Hi @daiwc , after i went thru [SORT](https://arxiv.org/pdf/1602.00763.pdf) paper i can say following.

khmot implements classical approaches for data association (Hungatian) and motion estimation (Kalman) and SORT also uses them. So far the difference that SORT also uses FastRCNN detector of people, but khmot assumes user provides coordinates of the objects (along with covariance matrix). Also SORT and khmot use different motion models (in Kalman filter). khmot uses `x, y, yaw,vel x, vel y, vel yaw`, but SORT uses `x, y, s, r, vel x, vel, y, vel s`, where `s` is scale and `r` aspect ratio (in paper they named `u, v, s, r, u' v', s'`). Also khmot supports non-omnidirectional motion, for example cars, but SORT does not. SORT assumes objects tracked on image, but khmot assumes object are reprojected to the ground plane (that's why it makes sense to use constrained motion models like non-omnidirecitonal).

In short, they use both same algorithms with different motion model. SORT is complete algorithm to detect and track pedestrians (bounding boxes) on images. khmot is only tracking for abstract objects, so detection should be done separately. khmot also do not take into account any object specifics except position and orientation, but SORT does (uses scale and aspect ration of bbox).

My main intention to create khmot was to use it for obstacles tracking that i get from my perception pipeline and i needed quite generic interface (e.g. i can pass only center coordinates).

Hope this helps.

**Comment by daiwc** [comment]
Thanks a lot！！ have you published paper about this algorithm?  have you test the algorithm on MOT benchmark?




------------------ 原始邮件 ------------------
发件人: "Roma Sokolkov"<notifications@github.com>; 
发送时间: 2019年10月15日(星期二) 上午8:46
收件人: "r7vme/khmot"<khmot@noreply.github.com>; 
抄送: "戴伟聪"<493980892@qq.com>; "Mention"<mention@noreply.github.com>; 
主题: Re: [r7vme/khmot] the different between khmot and sort? (#1)




Hi @daiwc , after i went thru SORT paper i can say following.
 
khmot implements classical approaches for data association (Hungatian) and motion estimation (Kalman) and SORT also uses them. So far the difference that SORT also uses FastRCNN detector of people, but khmot assumes user provides coordinates of the objects (along with covariance matrix). Also SORT and khmot use different motion models (in Kalman filter). khmot uses x, y, yaw,vel x, vel y, vel yaw, but SORT uses x, y, s, r, vel x, vel, y, vel s, where s is scale and r aspect ratio (in paper they named u, v, s, r, u' v', s'). Also khmot supports non-omnidirectional motion, for example cars, but SORT does not. SORT assumes objects tracked on image, but khmot assumes object are reprojected to the ground plane (that's why it makes sense to use constrained motion models like non-omnidirecitonal).
 
In short, they use both same algorithms with different motion model. SORT is complete algorithm to detect and track pedestrians (bounding boxes) on images. khmot is only tracking for abstract objects, so detection should be done separately. khmot also do not take into account any object specifics except position and orientation, but SORT does (uses scale and aspect ration of bbox).
 
My main intention to create khmot was to use it for obstacles tracking that i get from my perception pipeline and i needed quite generic interface (e.g. i can pass only center coordinates).
 
Hope this helps.

**Comment by r7vme** [comment]
> have you published paper about this algorithm?

noop, there is no innovation i did. All are basic state-of-the-art algorithms. Check out links in [credits section](https://github.com/r7vme/khmot#credits). Especially first one about data association. Hungarian mentioned there as good algorithm that finds global optimal solution in polynomial time. Regarding motion estimation check out second link about classical Kalman filter.

> have you test the algorithm on MOT benchmark?

Not really, it's definitely makes sense to benchmark it. In general, i need to plug in some detection and i do not expect it will perform any better than SORT. Because khmot does not take into account any object specifics (e.g. SORT uses scale and ratio of bounding boxes). May be check out later. Next step for me is to add ROS integration and check with my own data.

**Comment by r7vme** [comment]
i believe i answered the question. In the mean time i added ROS1 support and fixed tons of bugs in my code. Still a lot of fune tuning (hello Kalman filtering) required to get stuff running for your specific use case.
